### PR TITLE
Allow the 'params' keyword in indexers

### PR DESCRIPTION
**Pull request by sys27**
**Bugs this fixes:**

#17384 - Autocompletion list misses params for indexers

**Risk**

Low.

**Performance impact**

Low.

**Is this a regression from a previous update?**

No.

### CODE REVIEW

File: src/Workspaces/CSharp/Portable/Extensions/ContextQuery/SyntaxTreeExtensions.cs [hunk 1]
```diff
@@ -1042,6 +1042,28 @@ public static bool IsDefaultExpressionContext(this SyntaxTree syntaxTree, int po
...
+            this SyntaxTree syntaxTree,
+            int position,
+            SyntaxToken tokenOnLeftOfPosition,
+            CancellationToken cancellationToken)
+        {
+            var token = tokenOnLeftOfPosition;
+            token = token.GetPreviousTokenIfTouchingWord(position);
+
+            if (syntaxTree.IsParameterModifierContext(position, tokenOnLeftOfPosition, cancellationToken))
+                return true;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/Extensions/ContextQuery/SyntaxTreeExtensions.cs]: curlies always required.

File: src/Workspaces/CSharp/Portable/Extensions/ContextQuery/SyntaxTreeExtensions.cs [hunk 2]
```diff
@@ -1042,6 +1042,28 @@ public static bool IsDefaultExpressionContext(this SyntaxTree syntaxTree, int po
...
+
+            if ((token.IsKind(SyntaxKind.OpenBracketToken) || token.IsKind(SyntaxKind.CommaToken)) &&
+                token.Parent.IsKind(SyntaxKind.BracketedParameterList) &&
+                token.Parent.IsParentKind(SyntaxKind.IndexerDeclaration))
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/Extensions/ContextQuery/SyntaxTreeExtensions.cs]: please don't mix &&'s and ||'s in a single condition.  It gets difficult to read.  This can be written as:

```c#
if (token.IsKind(SyntaxKind.OpenBracketToken) || token.IsKind(SyntaxKind.CommaToken))
{
    return token.IsParentKind(SyntaxKind.BracketedParameterList);
}

return false;
```

File: src/Workspaces/CSharp/Portable/Extensions/ContextQuery/SyntaxTreeExtensions.cs (hunk 1 above)
Comment (sys27) [inline src/Workspaces/CSharp/Portable/Extensions/ContextQuery/SyntaxTreeExtensions.cs]: Ok.

File: src/Workspaces/CSharp/Portable/Extensions/ContextQuery/SyntaxTreeExtensions.cs (hunk 2 above)
Comment (sys27) [inline src/Workspaces/CSharp/Portable/Extensions/ContextQuery/SyntaxTreeExtensions.cs]: Why did you remove 'SyntaxKind.IndexerDeclaration'? Is 'SyntaxKind.BracketedParameterList' used only for indexers?

Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/Extensions/ContextQuery/SyntaxTreeExtensions.cs]: Exactly :)

File: src/Workspaces/CSharp/Portable/Extensions/ContextQuery/SyntaxTreeExtensions.cs [hunk 3]
```diff
@@ -1042,6 +1042,28 @@ public static bool IsDefaultExpressionContext(this SyntaxTree syntaxTree, int po
             return false;
         }
 
+        public static bool IsParamsModifierContext(
+            this SyntaxTree syntaxTree,
+            int position,
+            SyntaxToken tokenOnLeftOfPosition,
+            CancellationToken cancellationToken)
+        {
+            var token = tokenOnLeftOfPosition;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/Extensions/ContextQuery/SyntaxTreeExtensions.cs]: move these two lines below the first check.  They're not needed until the second check.

**Comment by CyrusNajmabadi** [comment]
Tagging @ManishJayaswal 

**Customer scenario**

User wants to type an indexer with a "params" parameter.  Because "params" is not in the completion list, they get tripped up.

**Bugs this fixes:**

https://github.com/dotnet/roslyn/issues/17384

**Workarounds, if any**

None.

**Risk**

Low.

**Performance impact**

Low.

**Is this a regression from a previous update?**

No/

**Root cause analysis:**

Misunderstanding of what was/wasn't allowed in C#.  Test case has been modified to reflect actual behavior of hte language.

**How was the bug found?**

Customer reported.

**Comment by ManishJayaswal** [comment]
Approved pending CR and tests passing

**Comment by CyrusNajmabadi** [comment]
Test failures are unrelated.  Merging in. Thanks for the contribution @sys27 !
